apps/emqx_plugins/src/emqx_plugins_serde.erl

```diff
@@ -0,0 +1,275 @@
+                        msg => "plugin_schema_add_failed",
+                        plugin => Name,
+                        reason => emqx_utils:readable_error_msg(Reason)
+                    }),
+                    E
+            end
+    end.
+
+get_schema(NameVsn) ->
+    Path = emqx_plugins:avsc_file_path(NameVsn),
+    case read_avsc_file(Path) of
+        {ok, Avsc} ->
+            {ok, Avsc};
+        {error, Reason} ->
+            ?SLOG(warning, Reason),
+            {error, Reason}
+    end.
+
+-spec delete_schema(schema_name()) -> ok | {error, term()}.
+delete_schema(NameVsn) ->
+    case lookup_serde(NameVsn) of
+        {ok, _Serde} ->
+            async_delete_serdes([NameVsn]),
+            ok;
+        {error, not_found} ->
+            {error, not_found}
+    end.
+
+-spec decode(schema_name(), encoded_data()) -> {ok, decoded_data()} | {error, any()}.
+decode(SerdeName, RawData) ->
+    with_serde(
+        "decode_avro_json",
+        eval_serde_fun(?FUNCTION_NAME, "bad_avro_binary", SerdeName, [RawData])
+    ).
+
+-spec encode(schema_name(), decoded_data()) -> {ok, encoded_data()} | {error, any()}.
+encode(SerdeName, Data) ->
+    with_serde(
+        "encode_avro_json",
+        eval_serde_fun(?FUNCTION_NAME, "bad_avro_data", SerdeName, [Data])
+    ).
+
+%%-------------------------------------------------------------------------------------------------
+%% `gen_server' API
+%%-------------------------------------------------------------------------------------------------
+
+init(_) ->
+    process_flag(trap_exit, true),
+    ok = emqx_utils_ets:new(?PLUGIN_SERDE_TAB, [
+        public, ordered_set, {keypos, #plugin_schema_serde.name}
+    ]),
+    State = #{},
+    SchemasMap = read_plugin_avsc(),
+    {ok, State, {continue, {build_serdes, SchemasMap}}}.
+
+handle_continue({build_serdes, SchemasMap}, State) ->
+    _ = build_serdes(SchemasMap),
+    {noreply, State}.
+
+handle_call({build_serdes, NameVsn, Avsc}, _From, State) ->
+    BuildRes = do_build_serde(NameVsn, Avsc),
+    {reply, BuildRes, State};
+handle_call(_Call, _From, State) ->
+    {reply, {error, unknown_call}, State}.
+
+handle_cast({delete_serdes, Names}, State) ->
+    lists:foreach(fun ensure_serde_absent/1, Names),
+    {noreply, State};
+handle_cast(_Cast, State) ->
+    {noreply, State}.
+
+terminate(_Reason, _State) ->
+    ok.
+
+%%-------------------------------------------------------------------------------------------------
+%% Internal fns
+%%-------------------------------------------------------------------------------------------------
+
+read_plugin_avsc() ->
+    Pattern = filename:join([emqx_plugins:install_dir(), "*", "config_schema.avsc"]),
+    lists:foldl(
+        fun(AvscPath, AccIn) ->
+            case read_avsc_file(AvscPath) of
+                {ok, Avsc} ->
+                    [_, NameVsn | _] = lists:reverse(filename:split(AvscPath)),
+                    AccIn#{to_bin(NameVsn) => Avsc};
```

Comment:
There is no need to keep the file content in process state ?
Read the file when building the serde.

Reply:
process `State` will always be empty map.
files contents will be dropped after building the serdes.

Reply:
right.